hail/src/main/scala/is/hail/lir/SplitMethod.scala

```diff
@@ -0,0 +1,204 @@
+          val size = 1 + x.children.map(visit).sum
+
+          if (size > SplitMethod.TargetMethodSize / 2) {
+            val l = m.newLocal("spill_large_expr", x.ti)
+            x.replace(load(l))
+            c.insertBefore(store(l, x))
+            1
+          } else
+            size
+        }
+
+        c.children.foreach(visit)
+      }
+
+      var x = b.first
+      while (x != null) {
+        splitLargeStatement(x)
+        x = x.next
+      }
+    }
+  }
+
+  def spillLocals(): Unit = {
+    val (locals, localIdx) = m.findAndIndexLocals(blocks)
+
+    val fields = locals.map { l =>
+      if (l.isInstanceOf[Parameter])
+        null
+      else
+        c.newField(genName("f", l.name), l.ti)
+    }
+
+    def localField(l: Local): Field =
+      l match {
+        case p: Parameter =>
+          if (p.i == 0)
+            null
+          else
+            paramFields(p.i - 1)
+        case _ => fields(localIdx(l))
+      }
+
+    def spill(x: X): Unit = {
+      x.children.foreach(spill)
+      x match {
+        case x: LoadX =>
+          val f = localField(x.l)
+          if (f != null)
+            x.replace(getField(f, load(m.getParam(0))))
+        case x: IincX =>
+          val f = localField(x.l)
+          assert(f != null)
+          x.replace(
+            putField(f, load(m.getParam(0)),
+              insn(IADD,
+                getField(f, load(m.getParam(0))),
+                ldcInsn(x.i, IntInfo))))
+        case x: StoreX =>
+          val f = localField(x.l)
+          assert(f != null)
+          val v = x.children(0)
+          v.remove()
+          x.replace(putField(f, load(m.getParam(0)), v))
+        case _ =>
+      }
+    }
+
+    for (b <- blocks) {
+      var x = b.first
+      while (x != null) {
+        val n = x.next
+        spill(x)
+        x = n
+      }
+    }
+  }
+
+  def splitBlock(b: Block): Unit = {
+    val last = b.last
+
+    val returnTI = last match {
+      case _: GotoX => UnitInfo
+      case _: IfX => BooleanInfo
+      case _: SwitchX => IntInfo
+      case _: ReturnX => m.returnTypeInfo
+      case _: ThrowX => UnitInfo
+    }
+
+    var L = new Block()
+    var x = b.first
+    var size = 0
+
+    while (x != last) {
+      if (size > SplitMethod.TargetMethodSize) {
+        val newM = c.newMethod(genName("m", "wrapped"), FastIndexedSeq.empty[TypeInfo[_]], UnitInfo)
+        L.method = newM
+        newM.setEntry(L)
+        L.append(returnx())
+
+        x.insertBefore(methodStmt(INVOKEVIRTUAL, newM, Array(load(m.getParam(0)))))
+
+        L = new Block()
+        size = 0
+      }
+
+      size += x.approxByteCodeSize()
+      val n = x.next
+      x.remove()
+      L.append(x)
+      x = n
+    }
+
+    val newM = c.newMethod(genName("m", "wrapped"), FastIndexedSeq.empty[TypeInfo[_]], returnTI)
+    L.method = newM
+    newM.setEntry(L)
+
+    def invokeNewM(): ValueX =
+      methodInsn(INVOKEVIRTUAL, newM, Array(load(m.getParam(0))))
```

Comment:
I'm not clear on the semantics of `ValueX`. From reading `SplitMethod`, it seems
* a `ValueX` may have arbitrary side effects,
* `X.children` must execute in exactly that order.

Is that right?

Reply:
1. Yes, `ValueX` can have side effects.  There is `MethodX` which represents a function call that returns a value.

2. I don't think we should rely on the order of execution of the children.  Let's call the order implementation defined (although in practice they're probably always executed in order.)

Reply:
What made me ask is this method:
```
      def splitLargeStatement(c: StmtX): Unit = {
        def visit(x: ValueX): Int = {
          // FIXME this doesn't handle many moderate-sized children
          val size = 1 + x.children.map(visit).sum

          if (size > SplitMethod.TargetMethodSize / 2) {
            val l = m.newLocal("spill_large_expr", x.ti)
            x.replace(load(l))
            c.insertBefore(store(l, x))
            1
          } else
            size
        }

        c.children.foreach(visit)
      }
```
For this to preserve the order of any side effects, doesn't it assume that the children of `c` are evaluated in the order of the `children` array? (And also that any side effects performed by `c` itself happen after the children are evaluated.)